app/code/Magento/Downloadable/Test/Unit/Observer/SetHasDownloadableProductsObserverTest.php

```diff
@@ -0,0 +1,184 @@
+     * @var CheckoutSession|MockObject
+     */
+    private $checkoutSessionMock;
+
+    /**
+     * @var SetHasDownloadableProductsObserver
+     */
+    private $setHasDownloadableProductsObserver;
+
+    /**
+     * @var Order|MockObject
+     */
+    private $orderMock;
+
+    /**
+     * Setup environment for test
+     */
+    protected function setUp()
+    {
+        $this->objectManager = new ObjectManager($this);
+
+        $this->orderMock = $this->createPartialMock(Order::class, ['getAllItems']);
+
+        $this->checkoutSessionMock = $this->createPartialMock(
+            CheckoutSession::class,
+            [
+                'getHasDownloadableProducts',
+                'setHasDownloadableProducts'
+            ]
+        );
+
+        $this->setHasDownloadableProductsObserver = $this->objectManager->getObject(
+            SetHasDownloadableProductsObserver::class,
+            [
+                'checkoutSession' => $this->checkoutSessionMock
+            ]
+        );
+    }
+
+    /**
+     * Test execute with session has downloadable products
+     */
+    public function testExecuteWithSessionHasDownloadableProducts()
+    {
+        $event = new DataObject(['item' => $this->orderMock]);
+        $observer = new Observer(['event' => $event]);
+
+        $this->checkoutSessionMock->method('getHasDownloadableProducts')->willReturn(true);
+        $this->orderMock->method('getAllItems')->willReturn([]);
+
+        $this->checkoutSessionMock->expects($this->never())
+            ->method('setHasDownloadableProducts')->with(true);
+
+        $this->setHasDownloadableProductsObserver->execute($observer);
+    }
+
+    /**
+     * Test execute with session has no downloadable products with the data provider
+     *
+     * @dataProvider executeWithSessionNoDownloadableProductsDataProvider
+     */
+    public function testExecuteWithSessionNoDownloadableProducts($allItems, $expectedCall)
+    {
+        $event = new DataObject(['order' => $this->orderMock]);
+        $observer = new Observer(['event' => $event]);
+
+        $allOrderItemsMock = [];
+        foreach ($allItems as $item) {
+            $allOrderItemsMock[] = $this->createOrderItem(...$item);
+        }
+
+        $this->checkoutSessionMock->method('getHasDownloadableProducts')->willReturn(false);
+
+        $this->orderMock->method('getAllItems')->willReturn($allOrderItemsMock);
+
+        $this->checkoutSessionMock->expects($expectedCall)
+            ->method('setHasDownloadableProducts')->with(true);
+
+        $this->setHasDownloadableProductsObserver->execute($observer);
+    }
+
+    /**
+     * Create Order Item Mock
+     *
+     * @param string $productType
+     * @param string $realProductType
+     * @param string $isDownloadable
+     * @return Item|MockObject
+     */
+    private function createOrderItem(
+        $productType = DownloadableProductType::TYPE_DOWNLOADABLE,
+        $realProductType = DownloadableProductType::TYPE_DOWNLOADABLE,
```

Comment:
I see that you always test with the same value for `productType and `realProductType`.
In that specific case, it's against YAGNI :-)

Removing the redundant argument, you'll also get side effect of cleaner DataProvider method (oneliners)